### PR TITLE
[🪛 CI improvements] fix mergify automerge behaviour

### PR DESCRIPTION
**Pull request by niccoloraspa**
Closes: #2747

## What is the purpose of the change

This PR implements the requested changes to improve the mergify config:

1. Waiting for 2 approvals before automerging.
2. Waiting for all CI jobs to be successful.
3. Waiting for 0 requested for changes. (I added this but it can be removed)

According to the documentation at https://docs.mergify.com/conditions/#github-actions, in order to wait for a GitHub action to succeed, the job name is used to identify the workflow.

As a result, I have changed the job name for all the required GitHub actions. During this process, I also took the opportunity to add documentation to our GitHub actions and improve the overall structure to ensure consistency.

While most of these changes are aesthetic rather than practical, it is important to verify that they do not break the purpose of the CI. Please review the changes to ensure their integrity and that the CI pipeline remains functional.

I have tested this in my local fork but I think it's best to test here before mergin, with a fake PR with the `A: automerge` label.

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [x] N/A

### CODE REVIEW

**Comment by niccoloraspa** [comment]
@ValarDragon we should check that this PR doesn't break the GitHub protection rules.

**Comment by mattverse** [comment]
@niccoloraspa wdym by "Waiting for 0 requested for changes. (I added this but it can be removed)"?

**Comment by p0mvn** [comment]
Hello @niniccoloraspa,

Thank you for your pull request and for working on improving the Mergify config and the overall structure of the GitHub actions.

Based on my review, the changes look appropriate, and I appreciate that you included documentation for each action. Below are some additional steps to proceed with:

1. Please make sure to apply the correct labels to this PR, such as `A: automerge`, `T:CI`, `V:state/compatible/no_backport`, and `A:no-changelog`.
2. Verify that all CI checks have passed successfully and there are no failures. If there are any issues, please address them before proceeding.
3. Once the above steps are completed, request a review by tagging @p0mvn.

After these steps, we can consider testing with a fake PR to ensure everything works as expected.

Overall, your changes look good, and I appreciate your effort in testing this on your local fork before submitting the PR.

**Comment by p0mvn** [comment]
Hello @niccoloraspa,

Thank you for submitting this pull request to improve the Mergify config and the overall structure of the GitHub actions.

Before we proceed, please kindly ensure the following:

1. Apply the correct labels to the pull request.
2. Make sure that CI is not failing, and all the checks have passed.
3. If necessary, add a changelog entry to the `Unreleased` section of `CHANGELOG.md`. If not needed, you can add the "A:no-changelog" label.
4. As you have mentioned "V:state-compatible/no_backport" label, please also add a label specifying the version to backport to in the format "A:backport/v".
5. Perform a self-review to double-check your changes.
6. Make sure that you have added relevant tests.

Once all of the above steps are complete, please tag @p0mvn for a manual review.

Please note that this message is automated.

**Comment by p0mvn** [comment]
Hello @niccoloraspa,

Thank you for submitting this pull request to improve the Mergify config and the overall structure of the GitHub actions.

Before we proceed, please kindly ensure the following:

1. Apply the correct labels to the pull request.
2. Make sure that CI is not failing, and all the checks have passed.
3. If necessary, add a changelog entry to the `Unreleased` section of `CHANGELOG.md`. If not needed, you can add the "A:no-changelog" label.
4. As you have mentioned "V:state-compatible/no_backport" label, please also add a label specifying the version to backport to in the format "A:backport/v".
5. Perform a self-review to double-check your changes.
6. Make sure that you have added relevant tests.

Once all of the above steps are complete, please tag @p0mvn for a manual review.

Please note that this message is automated.

**Comment by p0mvn** [comment]
@niccoloraspa please fix conflicts and let me know once R4R

**Comment by niccoloraspa** [comment]
@p0mvn I fixed the merge conflicts.

I see that a merging condition is not met. I think this is due the change of ids I had to do. 

Schermata 2023-08-22 alle 12 43 52


@ValarDragon is this an admin setting? Could you help me out?

**Comment by p0mvn** [comment]
Hi @niccoloraspa any updates on this? Would be great to have this in soon

**Comment by niccoloraspa** [comment]
Fix merged conflicts. 

I had to change some `ids` so the checks are currently hanging:
image

We might have to change some GitHub protection rules before merging.

File: .github/mergify.yml
Comment (p0mvn) [inline .github/mergify.yml:22]: Wasn't this renamed to `add-issue-to-project`?

File: .github/mergify.yml
Comment (p0mvn) [inline .github/mergify.yml:19]: Can we keep this at 1? Tbh this is mostly useful for simple PRs that don't require 2 approvals

File: .github/workflows/build.yml
Comment (p0mvn) [inline .github/workflows/build.yml:59]: Conflict

File: .github/mergify.yml
Comment (p0mvn) [inline .github/mergify.yml:29]: Does this exist?

File: .github/mergify.yml
Comment (p0mvn) [inline .github/mergify.yml:42]: Do these even run for regulat PRs?

File: .github/mergify.yml
Comment (p0mvn) [inline .github/mergify.yml:45]: Is this correct? Could not find another ref in PR with ctrl + f

**Review by p0mvn — CHANGES_REQUESTED** [review]
Thanks for refreshing this @niccoloraspa 

Had some comments, please take a look.

I'm wondering if new name IDs were updated everywhere across the repo where needed?

Also, was this tested on a personal fork? Would be great to try and see if it works before merging

File: .github/mergify.yml
Comment (niccoloraspa) [reply .github/mergify.yml:19]: sure

File: .github/workflows/build.yml
Comment (niccoloraspa) [reply .github/workflows/build.yml]: done. File didn't got saved :/

File: .github/mergify.yml
Comment (niccoloraspa) [reply .github/mergify.yml:42]: Only the dev one, but we can avoid it as a check

File: .github/mergify.yml
Comment (niccoloraspa) [reply .github/mergify.yml]: I changed to `e2e`. It was some time ago not sure if `tests-e2e` it's correct.
I will leave this open until it's tested

**Comment by niccoloraspa** [comment]
Addressed comments (thank you) and now testing in a fork

File: .github/mergify.yml
Comment (niccoloraspa) [reply .github/mergify.yml:29]: Yeah.

```yaml
  optimize-contracts:
    name: optimize-contract-${{ matrix.contract.name }}
```

where contract.name is:

```yaml
    strategy:
      matrix:
        contract: 
          - 
            name: rate-limiter
            workdir: ./x/ibc-rate-limit/
            bytecode: x/ibc-rate-limit/bytecode/rate_limiter.wasm
            build: artifacts/rate_limiter.wasm
          - 
            name: crosschain-swaps
            workdir: ./cosmwasm/
            bytecode: tests/ibc-hooks/bytecode/crosschain_swaps.wasm
            build: artifacts/crosschain_swaps.wasm
          - 
            name: swaprouter
            workdir: ./cosmwasm/
            bytecode: tests/ibc-hooks/bytecode/swaprouter.wasm
            build: artifacts/swaprouter.wasm
          - 
            name: outpost
            workdir: ./cosmwasm/
            bytecode: tests/ibc-hooks/bytecode/outpost.wasm
            build: artifacts/outpost.wasm
```

**Comment by niccoloraspa** [comment]
Successfully tested this in my fork: https://github.com/niccoloraspa/osmosis/pull/3

Mergify did automerge my PR

**Comment by niccoloraspa** [comment]
Only thing left is to change this ids in the branch protection rules:
image

**Review by p0mvn — APPROVED** [review]
LGTM

**Comment by p0mvn** [comment]
Thank you @niccoloraspa . Can we change the IDs in a separate PR to avoid blocking the rest of the changes?

**Pull request closed without merging**
